Add PromiseInput class which serves as input to Promise chain.

maffoo opened the pull request:
In several places, we were pulling out the resolve and reject callbacks from a newly-created promise and storing them to be called later when we have a value or error with which to complete the promise. This adds a `PromiseInput` class to formalize this pattern.

I'm not too sure about the name, so would be happy if people have other suggestions. I thought about `Completable` which conveys the notion that once we resolve or reject, the operation is complete and we can't resolve or reject again. But it's also nice to have the name reflect that this class is closely related to `Promise`. As always, names are hard.

DanielSank commented:
What's the need for the `_*` methods?

joshm-3b6b58 commented:
@maffoo it would be nice to start adding unit tests for every new piece. I don't know if that's necessary for this PR though since I merged the unit testing branch probably after you started this.

maffoo commented:
@DanielSank, `_resolve` and `_reject` are not methods, they are class members that happen to have function type. The methods `resolve` and `reject` just call the functions contained in those variables. I did it this way because public members of a class are mutable and I want users of this class to _only_ be able to call but not modify these functions, which methods do.

@joshmutus, good call; I'll add a couple of tests. If nothing else, they're helpful for documentation :)

maffoo commented:
After consultation with @DanielSank, renamed `PromiseInput` to `Obligation`. Also added basic tests for this class. @DanielSank and @joshmutus, please have another look when you can.

DanielSank commented:
@joshmutus if you want to review this together I'm game. It took me more than an hour (and more importantly a discussion with @maffoo) before I understood enough about the (terribly documented) ES6 Promise API such that I could even comprehend this code.

joshm-3b6b58 commented:
@DanielSank I'm in the cleanroom tomorrow, we can look at it together if it can wait until Friday.

DanielSank commented on client-js/app/scripts/obligation.ts:
```diff
@@ -0,0 +1,48 @@
+ */
+export class Obligation<R> {
+
+  public promise: Promise<R>;
```
Why public? It seems to me the only reason to make this public is so that an asynchronous API can do something like this:

``` javascript
somethingAsync(): Promise<A>() {
  var obligation = new Obligation<A>();
  myObligations.push(obligation);
  return obligation.promise;
}
```

If the only point of making the `promise` field public is to return it to the piece of code to which we're making the promise, I'd rather just use a Promise/Obligation factory and keep the `promise` field of `Obligation` private.

maffoo replied:
I tried converting `Obligation` to be an interface and using a factory function to produce `Obligation/Promise` pairs instead of having an `Obligation` hold a reference to the `Promise`: https://github.com/labrad/scalabrad-web/commit/2b79ab5ddeabc3e9afe302f7d032a981eea21a86

It's very javascripty (using an interface and structural types instead of classes), I have to say I don't like it as well. For me, just creating one `Obligation` object and having it hold a reference to the `Promise` is simpler. But take a look at that commit and let me know what you think.

DanielSank replied:
I have read and comprehended the code in https://github.com/labrad/scalabrad-web/commit/2b79ab5ddeabc3e9afe302f7d032a981eea21a86, and I like it much better. The original way of having the `Promise` live inside an `Obligation` is asymmetric for no reason. In the new style where we say

``` javascript
var {obligation, promise} = obligate<A>();
```

the symmetry makes sense; `obligate` and `promise` are on equal footing.

DanielSank replied:
I should point out that I know almost nothing about js and so am totally unaware of any implications that come from using an interface and "structural types" (don't even know what that means).

DanielSank commented on client-js/app/scripts/obligation.ts:
```diff
@@ -0,0 +1,48 @@
+   * Resolve our Promise with the given value.
+   */
+  resolve(value?: R | Promise<R>): void {
+    this._resolve(value);
```
~~Why not~~

``` javascript
this.promise.resolve(value);
```

~~which would also let us remove the mind-boggling constructor?~~

I see. Man the ES6 Promise API is really weird.

DanielSank commented on client-js/app/scripts/obligation.ts:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Obligation is the input side complement of Promise; if one party is promised
+ * a value, then of necessity another party must be obliged to provide the
```
I think "by necessity" is more standard. It might be more compact and clear to reword as

"If one party is promised a value, then the party who made the promise is obliged to provide that value or explain why they failed to do so".

maffoo replied:
"of necessity" = 26.5 M google results; "by necessity" = 18.8 M google results. Therefore "of necessity" wins :)

But yes, I like your wording better anyway.

DanielSank replied:
Eh, that's a 60% / 40% split. I'd say that's easily within the range where we can pick whichever we like better. Besides, most of those hits are probably middle-schoolers' Facebook posts :P

DanielSank commented on client-js/app/scripts/obligation.ts:
```diff
@@ -0,0 +1,48 @@
+ * available as arguments passed in to the executor function we pass to the
+ * Promise constructor. But in many situations we need to store the callbacks
+ * for later use or access them outside the constructor, which Obligation makes
+ * possible. This is especially useful in situations where we need to interface
```
The `resolve` and `reject` methods of the `Promise`living inside each `Obligation` are private. That being the case, how does `Obligation` make them accessible?

maffoo replied:
I don't understand what you are asking. `Promise` does not have methods called `resolve` and `reject`. The `Promise` _constructor_ takes a function which gets called with two functions to resolve or reject the promise, respectively. We store those functions in private members of `Obligation` but make them accessible by letting you call them through the public `resolve` and `reject` methods on `Obligation` itself.

DanielSank replied:
Can you just change "access them" to "call them"?

DanielSank replied:
Also, you say `Promise` has no `resolve` and `reject` methods, but then [Mozilla says otherwise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). Is there a reference you'd recommend that's not the completely unreadable official ES6 spec?

maffoo replied:
Those are not methods on a Promise instance, those are static methods that construct pre-resolved or pre-failed Promise instances. (They "lift a constant value into the Promise monad", if you like, though I'm guessing you don't.)

DanielSank replied:
> Those are not methods on a Promise instance, those are static methods that construct pre-resolved or pre-failed Promise instances.

Ah.

> They "lift a constant value into the Promise monad", if you like, though I'm guessing you don't.

Pretty sure last time I asked you what a monad was you said "everything".

DanielSank replied:
Anyway, I still think changing "access them" to "call them" would be helpful. Saying "access" makes me think I can get references to the `resolve` and `reject` functions and do whatever I want with them, which is not true. All I can do is call them through the Obligation's methods.

DanielSank commented:
Going with the "symmetric" version (i.e. the `obligate` function) LGTM.
